sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -129,11 +131,17 @@ class CodegenContext {
    *
    * They will be kept as member variables in generated classes like `SpecificProjection`.
    */
-  val mutableStates: mutable.ArrayBuffer[(String, String, String)] =
-    mutable.ArrayBuffer.empty[(String, String, String)]
+  private var mutableStates: mutable.ArrayBuffer[(String, String, String, Boolean)] =
+    mutable.ArrayBuffer.empty[(String, String, String, Boolean)]
 
-  def addMutableState(javaType: String, variableName: String, initCode: String): Unit = {
-    mutableStates += ((javaType, variableName, initCode))
+  def addMutableState(javaType: String, variableName: String,
+      initCode: String, isWholeStageResultVar: Boolean = false): Unit = {
+    if (isWholeStageResultVar) {
+      mutableStates = mutableStates.filterNot(state => (state._1 == javaType) && state._4)
+      mutableStates += ((javaType, variableName, initCode, isWholeStageResultVar))
```

Comment:
So you're effectively letting later-added states to override earlier-added ones that are equivalent and replaceable. Could that be done in a add-if-absent way instead?
Also, what's the criteria for a state to be "replaceable" in your design? Could you please make that explicit as comment in the code?